include/merlin/memory_pool.hpp

```diff
@@ -0,0 +1,598 @@
+/*
+ * Copyright (c) 2022, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cuda_runtime_api.h>
+#include <algorithm>
+#include <array>
+#include <functional>
+#include <iostream>
+#include <memory>
+#include <mutex>
+#include <thread>
+#include <vector>
+#include "debug.hpp"
+
+namespace nv {
+namespace merlin {
+
+/**
+ * Allocators are used by the memory pool (and maybe other classes) to create
+ * RAII complient containers for buffers allocated in different memory areas.
+ */
+template <class T, class Allocator>
+struct AllocatorBase {
+  using type = T;
+  using sync_unique_ptr = std::unique_ptr<type, Allocator>;
+  using async_unique_ptr = std::unique_ptr<type, std::function<void(type*)>>;
+  using shared_ptr = std::shared_ptr<type>;
+
+  inline static sync_unique_ptr make_unique(size_t n) {
+    return sync_unique_ptr(Allocator::alloc(n));
+  }
+
+  inline static async_unique_ptr make_unique(size_t n, cudaStream_t stream) {
+    return {Allocator::alloc(n, stream),
+            [stream](type* p) { Allocator::free(p); }};
+  }
+
+  inline static shared_ptr make_shared(size_t n, cudaStream_t stream = 0) {
+    return {Allocator::alloc(n, stream),
+            [stream](type* p) { Allocator::free(p, stream); }};
+  }
+
+  inline void operator()(type* ptr) { Allocator::free(ptr); }
+};
+
+/**
+ * Trivial fallback implementation using the standard C++ allocator. This mostly
+ * exists to ensure interface correctness, and as an illustration of what a
+ * proper allocator implementation should look like.
+ */
+template <class T>
+struct StandardAllocator final : AllocatorBase<T, StandardAllocator<T>> {
+  using type = typename AllocatorBase<T, StandardAllocator<T>>::type;
+
+  static constexpr const char* name{"StandardAllocator"};
+
+  inline static type* alloc(size_t n, cudaStream_t stream = 0) {
+    return new type[n];
+  }
+
+  inline static void free(type* ptr, cudaStream_t stream = 0) { delete[] ptr; }
+};
+
+/**
+ * Claim/release buffers in pinned host memory.
+ */
+template <class T>
+struct HostAllocator final : AllocatorBase<T, HostAllocator<T>> {
+  using type = typename AllocatorBase<T, HostAllocator<T>>::type;
+
+  static constexpr const char* name{"HostAllocator"};
+
+  inline static type* alloc(size_t n, cudaStream_t stream = 0) {
+    void* ptr;
+    CUDA_CHECK(cudaMallocHost(&ptr, sizeof(T) * n));
```

Comment:
If we use the CUDA API, Change the name to *.cuh, that will decrease the requirement on the compilation environment.

Reply:
Done!